### PR TITLE
test: fix code coverage

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

To have consistent and valid code coverage data. 

**Description**

- Fix the broken coverage upload
- All the packages were uploaded with same root dir as `src` and overriding the coverage data of other packages. 
- Remove unnecessary coverage flags per package, we can always do that by navigating to package directory. 
- Add a coverage flag `unit` to keep it separate, in case we introduce `integration` coverage in future. 

**Steps to test or reproduce**

- Run all the tests.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -194,7 +194,12 @@ jobs:
       #   if: ${{ failure() && steps.unit_tests.conclusion == 'failure' }}
 
       - name: Upload coverage data
-        run: yarn coverage
```
Comment (nazarhussain) [inline .github/workflows/test.yml:197]: These tasks are not more valid as npm package `codecov` is deprecated. And they recommend to use new native binaries instead. 
https://about.codecov.io/blog/introducing-codecovs-new-uploader/

File: package.json [hunk 2]
```diff
@@ -54,7 +53,6 @@
     "@typescript-eslint/parser": "6.7.2",
     "@vitest/coverage-v8": "^1.2.1",
     "@vitest/browser": "^1.2.1",
-    "codecov": "^3.8.3",
```
Comment (nazarhussain) [inline package.json:57]: This package is deprecated. https://about.codecov.io/blog/codecov-uploader-deprecation-plan/

**Review by jeluard — COMMENTED** [review]
GitHub actions could be kept up-to-date using [dependabot](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/keeping-your-actions-up-to-date-with-dependabot) but looks like it's not use on purpose?

**Comment by nazarhussain** [comment]
> GitHub actions could be kept up-to-date using [dependabot](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/keeping-your-actions-up-to-date-with-dependabot) but looks like it's not use on purpose?

Yes we don't like to use some bot contributions. 😄

**Comment by jeluard** [comment]
> > GitHub actions could be kept up-to-date using [dependabot](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/keeping-your-actions-up-to-date-with-dependabot) but looks like it's not use on purpose?
> 
> Yes we don't like to use some bot contributions. 😄

Makes sense, although they could be restricted to GitHub actions only.

File: .codecov.yml [hunk 3]
```diff
@@ -34,3 +34,9 @@ ignore:
   - "supporting-docs"
   - "docker"
   - ".github"
+
+flags:
+  unit:
+    paths:
+      - ".*"
```
Comment (jeluard) [inline .codecov.yml:41]: Maybe worth documenting why this is needed?

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -194,7 +194,12 @@ jobs:
...
+        uses: codecov/codecov-action@v4
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          fail_ci_if_error: true
+          verbose: true # optional (default = false)
+          flags: "unit"
```
Comment (jeluard) [inline .github/workflows/test.yml:202]: `unit` refers to the `.codecov.yml` right? Maybe worth commenting that?

File: README.md [hunk 5]
```diff
@@ -8,6 +8,7 @@
 [![Ethereum Consensus Spec v1.1.10](https://img.shields.io/badge/ETH%20consensus--spec-1.1.10-blue)](https://github.com/ethereum/consensus-specs/releases/tag/v1.1.10)
 ![ES Version](https://img.shields.io/badge/ES-2021-yellow)
 ![Node Version](https://img.shields.io/badge/node-20.x-green)
+[![codecov](https://codecov.io/gh/ChainSafe/lodestar/graph/badge.svg?token=kHAUjtLsDG)](https://codecov.io/gh/ChainSafe/lodestar)
```
Comment (jeluard) [inline README.md]: Is the token required? Looks like it works without.

Comment (nflaig) [inline README.md]: the badges look a bit overloaded, we have a second line now, might wanna think about about removing less relevant ones, e.g. do we really need the github release ref?

image

Comment (nazarhussain) [reply README.md]: How about having those in actual two lines. 
image

Comment (nflaig) [reply README.md]: definitely much better

Comment (nazarhussain) [reply README.md]: I copied the markdown directly from their interface, but for public repo don't need the token.

File: .github/workflows/test.yml (hunk 4 above)
Comment (nazarhussain) [reply .github/workflows/test.yml:202]: Yes `unit` refers to the `flag` defined in the `.codecov.yml` file.

File: .codecov.yml (hunk 3 above)
Comment (nazarhussain) [reply .codecov.yml:41]: Added into PR description already. It's very common to practice in `codecov` to flag the coverage into unit and integration or other tests. So I didn't found it reasonable to add some comment here.

**Review by jeluard — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.16.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.16.0) :tada:

**Pull request merged**